cg/meta/store/base.py

```diff
@@ -29,26 +35,110 @@ def add_new_analysis(bundle_data, case_obj, status, version_obj):
     ]
 
-    return data
 
+def _get_files_index(deliverables: dict, pipeline: list) -> list:
+    """ Get all index files from the deliverables file """
 
-def get_tags(file: dict, pipeline: str) -> list:
+    return [
+        {"path": file["path_index"], "tags": get_tags(file, pipeline), "archive": False,}
+        for file in deliverables["files"]
+        if file["path_index"]
+    ]
+
+
+def get_tags(file: dict, pipeline_tags: list) -> list:
     """Get all tags for a file"""
 
-    all_tags = [file["format"], file["id"], file["step"], file["tag"], pipeline]
+    all_tags = [file["format"], file["id"], file["step"], file["tag"]]
+    all_tags.extend(pipeline_tags)
     unique_tags = set(all_tags)
     only_existing_tags = unique_tags - set([None])
     sorted_tags = sorted(list(only_existing_tags))
 
     return sorted_tags
+
+
+def build_bundle(config_data: dict, analysisinfo_data: dict, deliverables: dict) -> dict:
+    """Create a new bundle to store in Housekeeper"""
+
+    pipeline = config_data["samples"][0]["type"]
+    pipeline_tag = HK_TAGS[pipeline]
+
+    data = {
+        "name": config_data["case"],
+        "created": analysisinfo_data["date"],
+        "pipeline_version": analysisinfo_data["version"],
+        "files": get_files(deliverables, pipeline_tag),
+    }
+    return data
+
+
+def _convert_tags(data: list):
+    """ Convert tags from deliverables tags to MIP standard tags """
+
+    for deliverables_tags, mip_tags in MIP_TAGS.items():
+        for file in data:
+            if all(tag in file["tags"] for tag in deliverables_tags):
+                tags_filtered = list(filter(lambda x: x not in deliverables_tags, file["tags"]))
+                converted_tags = tags_filtered + mip_tags["tags"]
+                file["tags"] = converted_tags
+
+
+def _check_mandatory_tags(files: list):
+    """
+        Check if all the mandatory tags are present for the files to be added to Housekeeper.
+        Raise an exception if not.
+    """
+
+    all_deliverable_tags = [file["tags"] for file in files]
+    all_mandatory_tags = [tag for tag in MIP_TAGS if MIP_TAGS[tag]["is_mandatory"]]
+    deliverable_tags = _flatten(all_deliverable_tags)
+    mandatory_tags = _flatten(all_mandatory_tags)
+
+    tags_are_missing, missing_tags = _determine_missing_files(mandatory_tags, deliverable_tags)
+
+    if tags_are_missing:
+        raise MandatoryFilesMissing(
+            f"Mandatory files are missing! These are the missing tags: {missing_tags}."
+        )
+
+
+def _determine_missing_files(mandatory_tags: set, found_tags: set) -> tuple:
+    """Determines if mandatory are missing, and which ones"""
+
+    missing_tags = mandatory_tags.difference(found_tags)
+    are_tags_missing = bool(len(missing_tags) > 0)
+
+    return are_tags_missing, missing_tags
+
+
+def _flatten(nested_list: list) -> set:
+    """ Flattens a nested tag list and returns all unique tags"""
+
+    return {tag for tag_list in nested_list for tag in tag_list}
```

Comment:
Could we break this one up a bit? It requires a bit of think to understand what it does. Double for loops in one line is a bit much for my brain sometimes.